base/broadcast.jl

```diff
@@ -219,6 +219,18 @@ argtype(bc::Broadcasted) = argtype(typeof(bc))
 _eachindex(t::Tuple{Any}) = t[1]
 _eachindex(t::Tuple) = CartesianIndices(t)
 
+_combined_indexstyle(::Type{Tuple{A}}) where A = IndexStyle(A)
+_combined_indexstyle(::Type{TT}) where TT =
+    IndexStyle(IndexStyle(Base.tuple_type_head(TT)),
+               _combined_indexstyle(Base.tuple_type_tail(TT)))
+
+Base.IndexStyle(bc::Broadcasted) = IndexStyle(typeof(bc))
+Base.IndexStyle(::Type{<:Broadcasted{<:Any,Nothing,<:Any,Args}}) where {Args <: Tuple} =
+    _combined_indexstyle(Args)
+Base.IndexStyle(::Type{<:Broadcasted{<:Any}}) = IndexCartesian()
+
+Base.LinearIndices(bc::Broadcasted) = LinearIndices(eachindex(bc))
```

Comment:
I need to look into how indexing is defined for `Broadcasted` (especially for how it interacts with `Axes`) and if this `IndexStyle` is correct.

However, it seems to me that it'd be much simpler if we define `IndexStyle(::Broadcasted)` than `IndexStyle(::Type{<:Broadcasted})` as it would let us use `eachindex(::Broadcasted)` in `IndexStyle`.  Does it makes sense to define `IndexStyle(::Broadcasted)` directly than the type trait?

Reply:
AFAIK `IndexStyle` needs to take a type rather than an instance so that you can know statically what kind of indices to expect (that's more or less the definition of a trait). At least that's how its [documented](https://docs.julialang.org/en/latest/manual/interfaces/#man-interface-array-1).

Reply:
My comment is based on the implementation detail in reduce.jl that `IndexStyle` is called on _instance_

https://github.com/JuliaLang/julia/blob/236c69b1a53acf4057de830979138765c79aa596/base/reduce.jl#L299

which is valid since `IndexStyle` is callable on `AbstractArray`:

https://github.com/JuliaLang/julia/blob/236c69b1a53acf4057de830979138765c79aa596/base/indices.jl#L70

However, to make `IndexStyle` a proper trait for `Broadcasted`, I defined methods for _types_ and added the method for instance on top of them, just like how it's done for `AbstractArray`:

https://github.com/JuliaLang/julia/blob/f4c68eeeeb14bfd03a0d53ea011ebe0e05b32037/base/broadcast.jl#L227

Note that direct definition of `IndexStyle(::Broadcasted)` would be type-stable if `eachindex` is so.  If `eachindex` is not type-stable we don't have the performance gain anyway.  Initially, I wasn't aiming for adding public interface for `Broadcasted` in this PR so I thought relying on the implementation detail that `IndexStyle` is only called on instance was OK.  However, I also understand that contaminating "signature space" of `IndexStyle` is not particularly a clean solution.

If we say that `IndexStyle` for `Broadcasted` has to be a proper trait, we need to note that `IndexStyle` must be properly defined whenever `axes` is customized since [`axes` is a public interface](https://docs.julialang.org/en/latest/manual/interfaces/#man-interfaces-broadcasting-1).

Reply:
Yeah, but as long as you define `IndexStyle` it should be a proper trait, independent from how it's used internally. Otherwise, you'd better define a custom internal function.

Anyway, what's the problem with the current implementation in this PR? Just that's it's complicated?

Reply:
OK. So, reading the documentation and the code more, it looks like `Base.axes(::Broadcasted{MyStyle})` is not an overloadable method?  From the documentation, `Base.axes(x)` is overloadable if `x` participate in the broadcasting.  This method is then called via `combine_axes` in `axes(::Broadcasted)`.  I was initially worried that `IndexStyle(::Type{<:Broadcasted})` and `Base.axes(::Broadcasted)` can become incompatible if downstream package authors are not careful but it looks like I don't need to worry about it.